ecdsa/src/lib.rs

```diff
@@ -176,38 +186,36 @@ where
 {
     /// Get the `r` component of this signature
     pub fn r(&self) -> NonZeroScalar<C> {
-        let r_bytes = GenericArray::clone_from_slice(&self.bytes[..C::UInt::BYTE_SIZE]);
-        NonZeroScalar::from_repr(r_bytes)
+        NonZeroScalar::from_repr(self.split_bytes().0)
             .unwrap_or_else(|| unreachable!("r-component ensured valid in constructor"))
     }
 
     /// Get the `s` component of this signature
     pub fn s(&self) -> NonZeroScalar<C> {
-        let s_bytes = GenericArray::clone_from_slice(&self.bytes[C::UInt::BYTE_SIZE..]);
-        NonZeroScalar::from_repr(s_bytes)
-            .unwrap_or_else(|| unreachable!("r-component ensured valid in constructor"))
+        NonZeroScalar::from_repr(self.split_bytes().1)
+            .unwrap_or_else(|| unreachable!("s-component ensured valid in constructor"))
+    }
+
+    /// Split the signature into its `r` and `s` scalars.
+    pub fn split_scalars(&self) -> (NonZeroScalar<C>, NonZeroScalar<C>) {
+        (self.r(), self.s())
     }
 
     /// Normalize signature into "low S" form as described in
     /// [BIP 0062: Dealing with Malleability][1].
     ///
     /// [1]: https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki
-    pub fn normalize_s(&mut self) -> Result<bool>
+    pub fn normalize_s(&mut self) -> bool
     where
         Scalar<C>: NormalizeLow,
     {
-        let s_bytes = GenericArray::from_mut_slice(&mut self.bytes[C::UInt::BYTE_SIZE..]);
-        Scalar::<C>::from_repr(s_bytes.clone())
-            .map(|s| {
-                let (s_low, was_high) = s.normalize_low();
-
-                if was_high {
-                    s_bytes.copy_from_slice(&s_low.to_repr());
-                }
-
-                was_high
-            })
-            .ok_or_else(Error::new)
+        let (s_low, was_high) = self.s().normalize_low();
+
+        if was_high {
+            self.bytes[C::UInt::BYTE_SIZE..].copy_from_slice(&s_low.to_repr());
+        }
+
+        was_high
```

Comment:
Do I understand this diff correctly that `normalize_s` cannnot error anymore? Was there a possible error case before or did we use the `Result` only to make the type system happy?

Reply:
Previously `ecdsa::Signature` was a naive wrapper around a 64-byte array which could contain any value.

[The bytes are now validated by the constructor](https://github.com/RustCrypto/signatures/blob/6ddc647172c76b4ead0b2ed4d2e37c64812348e9/ecdsa/src/lib.rs#L280-L304) and guaranteed to always represent the big endian serialization of a valid nonzero element of the scalar field. Even better: this now works for any elliptic curve regardless of if it has a field arithmetic implementation by using a common implementation in the `elliptic-curve` crate.

That eliminates the error case that previously existed here, where it was possible that the bytes would not decode successfully as `NonZeroScalar`.

Reply:
Great, thanks a lot for the detailed answer. Much appreciated.